### PR TITLE
feat: add music to homepage and matchmaking screen

### PR DESCRIPTION
**Pull request by privilegemendes**
## Description
This PR introduces updates to the sound provider including a game loader to pre load all music files prior rendering the DOM.

# Changes
- Added all music files in an object with assigned trackId for ease of use
- Added home page and match-making music
- updated play music hooks to control when music is started and stopped based on game location

### CODE REVIEW

File: src/components/audio-settings/volume-slider.tsx [hunk 1]
```diff
@@ -5,24 +5,23 @@ import { AUDIO_OFF, AUDIO_ON } from "~/constants/index.js";
...
-    }
+    changeVolume(newValue as number);
   };
 
   return (
     <Box sx={{ width: 400 }}>
       <Stack spacing={2} direction="row" sx={{ mb: 1 }} alignItems="center">
         <Button variant="text" onClick={() => changeVolume(AUDIO_OFF)}>
-          <VolumeDown />
+          <VolumeDown color={disabled ? "disabled" : "primary"} />
```
Comment (Mautjee) [inline src/components/audio-settings/volume-slider.tsx:20]: Define this constant in a variable in the component.

File: src/components/audio-settings/volume-slider.tsx [hunk 2]
```diff
@@ -31,7 +30,7 @@ export const VolumeSlider: FC<Props> = ({ volume, changeVolume }) => {
           step={0.1}
         />
         <Button variant="text" onClick={() => changeVolume(AUDIO_ON)}>
-          <VolumeUp />
+          <VolumeUp color={disabled ? "disabled" : "primary"} />
```
Comment (Mautjee) [inline src/components/audio-settings/volume-slider.tsx:33]: Use the variable of before

File: src/containers/sound-provider/sound-provider.tsx [hunk 3]
```diff
@@ -51,13 +64,32 @@ export const SoundProvider: FC<Props> = ({ children }) => {
...
+    if (mainContainerRef.current === null) return;
+    const mainContainer = mainContainerRef.current;
+
+    mainContainer.addEventListener("click", () => {
+      if (audioContext && audioContext.state === "suspended") {
+        void audioContext.resume();
+      }
+    });
+
+    return () => {
```
Comment (ramidecodes) [inline src/containers/sound-provider/sound-provider.tsx:80]: Don't you need to remove the event listener on the return from an useEffect?

File: src/hooks/sounds.ts [hunk 4]
```diff
@@ -1,78 +1,97 @@
...
+    useAndRequireContext(SoundContextRef, "usePlayMusic", "sound-provider");
+
+  const sourceNodeRef = useRef<AudioBufferSourceNode | null>(null);
+  const router = useRouter();
+
+  useEffect(() => {
+    const playAudio = () => {
+      if (!audioContext || !masterGainNode || !musicGainNode) return;
+      const audioBuffer = audioBuffers.current.get(audioFile);
+      if (audioBuffer) {
```
Comment (ramidecodes) [inline src/hooks/sounds.ts:28]: Add an empty line before this conditional

File: src/hooks/sounds.ts [hunk 5]
```diff
@@ -1,78 +1,97 @@
...
+        sourceNodeRef.current = sourceNode;
       }
-    },
-    [audioContext, masterGainNode, musicGainNode],
-  );
+    };
+
+    // Function to stop the music
+    const stopAudio = () => {
+      if (sourceNodeRef.current) {
```
Comment (ramidecodes) [inline src/hooks/sounds.ts:46]: Would a conditional chain be enough for this case? 
` sourceNodeRef.current?.stop()`

File: src/pages/lobby.tsx [hunk 6]
```diff
@@ -31,17 +33,27 @@ const Lobby: FC = () => {
 
   api.lobby.onReadyToPlay.useSubscription(undefined, {
     onData({ roomId }) {
-      void push({ pathname: pages.match, query: { roomId } });
+      setTrack("MatchMakingOutro");
+      setTimeout(() => {
+        void push({ pathname: pages.match, query: { roomId } });
+      }, 2000);
```
Comment (ramidecodes) [inline src/pages/lobby.tsx:39]: Is this time related to any other existing constant? Should this be a constant?

**Review by Mautjee — APPROVED** [review]

**Review by ramidecodes — CHANGES_REQUESTED** [review]

**Review by ramidecodes — APPROVED** [review]

File: src/containers/game-loader/game-loader.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+      }
+    };
+
+    if (audioContext) {
+      void loadMusicTracks(trackEntries);
+    }
+  }, [audioBuffers, audioContext, trackEntries]);
+
+  return <>{children}</>;
+};
```
Comment (silimarius) [inline src/containers/game-loader/game-loader.tsx:55]: Why not do this in the audio context provider? Seems to me like this kind of initialisation should go there

File: src/hooks/volume.ts [hunk 8]
```diff
@@ -61,10 +69,14 @@ export const useChangeSFXVolume = () => {
...
-    useAndRequireContext(ContextRef, "useChangeMusicVolume", "sound-provider");
+    useAndRequireContext(
+      SoundContextRef,
+      "useChangeMusicVolume",
+      "sound-provider",
+    );
 
   useEffect(() => {
-    if (!isClient() || !masterGainNode || !musicGainNode) {
+    if (!masterGainNode || !musicGainNode) {
```
Comment (silimarius) [inline src/hooks/volume.ts:79]: What was wrong with the `!isClient()` check?

File: src/pages/_app.tsx [hunk 9]
```diff
@@ -7,15 +7,15 @@ import { WalletModalProvider } from "@demox-labs/aleo-wallet-adapter-reactui";
...
 import { api } from "~/utils/api.js";
 import { ThemeProvider } from "~/styles/index.js";
 import { APP_NAME } from "~/constants/index.js";
 import { useRouter } from "next/router.js";
 import { pages } from "~/router.js";
 import { AppContainer } from "~/containers/app-container/index.js";
 import { SoundProvider } from "~/containers/sound-provider/index.js";
-
-import "~/styles/globals.css";
+import { GameLoader } from "~/containers/game-loader";
```
Comment (silimarius) [inline src/pages/_app.tsx:18]: Extension, but actually just move the stuff from this component directly to the `SoundProvider`

File: src/pages/lobby.tsx [hunk 10]
```diff
@@ -1,17 +1,19 @@
 import { type FC, useState } from "react";
+import { type TrackId } from "~/constants/sounds.js";
 import { useRouter } from "next/router.js";
+import { usePlayMusic } from "~/hooks/sounds.js";
```
Comment (silimarius) [inline src/pages/lobby.tsx:4]: Import order

File: src/pages/lobby.tsx [hunk 11]
```diff
@@ -31,17 +34,27 @@ const Lobby: FC = () => {
 
   api.lobby.onReadyToPlay.useSubscription(undefined, {
     onData({ roomId }) {
-      void push({ pathname: pages.match, query: { roomId } });
+      setTrack("MatchMakingOutro");
```
Comment (silimarius) [inline src/pages/lobby.tsx:37]: Why not just use a handler that just plays the sound instead of setting a variable that then changes a value passed to a hook that plays the sound?

Comment (silimarius) [reply src/pages/lobby.tsx:37]: Same way `playSfx` works

**Review by silimarius — CHANGES_REQUESTED** [review]
